wallet-api/src/Wallet/Emulator/Types.hs

```diff
@@ -0,0 +1,138 @@
+{-# LANGUAGE ConstraintKinds   #-}
+{-# LANGUAGE FlexibleContexts  #-}
+{-# LANGUAGE GADTs             #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+module Wallet.Emulator.Types where
+
+import           Control.Monad.Except
+import           Control.Monad.Operational as Op
+import           Control.Monad.State
+import           Control.Monad.Writer
+import           Data.Either
+import           Data.Map                  as Map
+import           Data.Maybe
+import           Data.Text                 as T
+
+-- Basic types (should be replaced by "real" types later)
+
+-- don’t care what these are, some data to allow typeclasses
+data Txn = Txn Int
+    deriving (Show, Eq, Ord)
+-- each agent has its own wallet state
+data Agent = Agent Int
+    deriving (Show, Eq, Ord)
+
+type Block = [Txn]
+type Chain = [Block]
+type TxPool = [Txn]
+
+data Notification = BlockValidated Block
+                  | BlockHeight Int
+                  deriving (Show, Eq, Ord)
+
+-- Wallet code
+
+data WalletState = WalletState
+    deriving (Show, Eq, Ord)
+
+emptyWalletState :: WalletState
+emptyWalletState = WalletState
+
+-- manually records the list of transactions to be submitted
+newtype EmulatedWalletApi a = EmulatedWalletApi { runEmulatedWalletApi :: StateT WalletState (Writer [Txn]) a }
+    deriving (Functor, Applicative, Monad, MonadState WalletState, MonadWriter [Txn])
+
+-- TODO
+instance WalletAPI EmulatedWalletApi where
+    submitTxn txn = tell [txn]
+    handleNotifications = undefined
+
+-- TODO
+class WalletAPI m where
+    submitTxn :: Txn -> m ()
+    handleNotifications :: [Notification] -> m ()
+
+data FailureReason = FailureReason T.Text
+    deriving Show
+
+-- Emulator code
+
+-- This is so that assertions can tell whether and why things failed. They’re not visible through the wallet api, but only to assertions, which is helpful for testing
+data FailedTxn = FailedTxn Txn FailureReason
+    deriving Show
+type FailedTxPool = [FailedTxn]
+
+type Assertion = EmulatorState -> Maybe AssertionError
+data AssertionError = AssertionError T.Text
+    deriving Show
+
+isValidated :: Txn -> Assertion
+isValidated txn emState =
+    if notElem txn (join $ emChain emState) then Just $ AssertionError $ "Txn not validated: " <> T.pack (show txn) else Nothing
+
+-- | The type of events in the emulator. @n@ is the type (usually a monad) in which wallet actions
+-- take place.
+data Event n a where
+    -- these need the coordination code to tell us what to do
+    WalletAction :: Agent -> n () -> Event n [Txn]
```

Comment:
I think we should add another parameter `m` for the wallet interactions that the client code is allowed to do. Or maybe even 
```
WalletAction :: forall m. WalletAPI m => Agent -> m () -> Event n [Txn]
```

By constraining it like that, and removing `handleNotifications` from `WalletAPI` we ensure that any code we write as a `WalletAction` can be run against the real wallet later on.

Reply:
That's exactly what `n` is, no? Just without the class constraint here.

Reply:
> That's exactly what n is, no? 

Yes, and in this place `n` is not right. Currently I can write

```
trace :: Trace ()
trace = Op.singleton $ WalletAction (Agent 1) (handleNotifications [BlockHeight 100])
```

Which does not make sense to me. And the two changes I proposed would ensure that this doesn't type check anymore (there are other changes you can make to achieve the same goal, which I would also be happy with :) )

The reason why I think it shouldn't type check is that inside `WalletAction` we want to write code that is agnostic of the wallet API implementation, so it could be executed against a real wallet also. And in that context `handleNotifications` does not make sense because the real wallet gets its blockchain notifications from a code node, not from the plutus client.

Reply:
Yes I 100% agree that `handleNotifications` should not be in `WalletApi`, but I don't think we need to change `Event`.